projects/ion/src/lib/tour/tour.service.ts

```diff
@@ -0,0 +1,203 @@
+import { DOCUMENT } from '@angular/common';
+import {
+  ApplicationRef,
+  ComponentFactoryResolver,
+  ComponentRef,
+  Inject,
+  Injectable,
+  Injector,
+} from '@angular/core';
+import { isEmpty, isEqual, isNull } from 'lodash';
+import { BehaviorSubject, Observable, Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+
+import { IonStartTourProps, IonTourStepProps } from '../core/types/tour';
+import { SafeAny } from '../utils/safe-any';
+import { IonTourBackdropComponent } from './tour-backdrop';
+
+type StepsMap = Map<IonTourStepProps['ionStepId'], IonTourStepProps>;
+
+@Injectable()
+export class IonTourService {
+  public currentStep = new BehaviorSubject<IonTourStepProps | null>(null);
+  public activeTour = new BehaviorSubject<string | null>(null);
+
+  private _tours: Record<string, StepsMap> = {};
+  private backdropRef: ComponentRef<IonTourBackdropComponent> | null = null;
+  private destroyBackdrop$ = new Subject<void>();
+
+  public get currentStep$(): Observable<IonTourStepProps | null> {
+    return this.currentStep.asObservable();
+  }
+
+  public get activeTour$(): Observable<string | null> {
+    return this.activeTour.asObservable();
+  }
+
+  public get steps(): IonTourStepProps[] {
+    return Array.from(this._tours[this.activeTourId].values());
+  }
+
+  private get activeTourId(): string {
+    return this.activeTour.value || Object.keys(this._tours)[0];
+  }
+
+  constructor(
+    @Inject(DOCUMENT) private document: SafeAny,
+    private componentFactoryResolver: ComponentFactoryResolver,
+    private injector: Injector,
+    private appRef: ApplicationRef
+  ) {}
+
+  public saveStep(step: IonTourStepProps): void {
+    if (!this._tours[step.ionTourId]) {
+      this._tours[step.ionTourId] = new Map();
+    }
+
+    const current = this.currentStep.value;
+
+    if (
+      current &&
+      current.ionStepId === step.ionStepId &&
+      !isEqual(step.target.toJSON(), current.target.toJSON())
+    ) {
+      this.navigateToStep(step);
+    }
+
+    this._tours[step.ionTourId].set(step.ionStepId, step);
+  }
+
+  public removeStep(stepId: IonTourStepProps['ionStepId']): void {
+    this._tours[this.activeTourId].delete(stepId);
+  }
+
+  public start(props: IonStartTourProps = {}): void {
+    setTimeout(() => {
+      if (isEmpty(this._tours)) {
+        console.error('No steps found!');
```

Comment:
Disable linter on this line if the console it's really intentional (same for consoles below).